### PR TITLE
Feat/optimise consensus sigcheck

### PR DESCRIPTION
**Pull request by AdoAdoAdo**
## Description of the reasoning behind the pull request
This PR brings several changes on the way block signatures are handler during consensus in order to optimize the CPU usage.
Each signature share used in the aggregation of signatures for the block consensus proof, was previously individually verified.
Assuming all validators are honest, the verification takes ~3ms per signature. In metachain, each block accumulates ~400 signatures which is individually checked by all participants in consensus, taking a cumulative ~1.2s CPU time for each block that passes consensus.

## Proposed Changes
This 1.2 seconds for metachain and ~200 ms for shards each round (for the synchronized nodes) can be optimized out, if instead the aggregation is done assuming all signature shares are valid, and verification is done on the aggregation instead.
The cost of the verification of the aggregated signature was optimized by the previous PR #4314 down to ~3ms as well.
This however can only be saved if all aggregated signatures are valid, otherwise, if the aggregated signature does not verify we will have an extra cost of 3ms, as the leader will aggregate first the signatures and then verify the aggregated signature. If the aggregated signature does not verify, the leader will have to check the signatures individually. 
In order to avoid the increase and instead always have this optimization, a penalty for providing invalid signatures will be implemented which we call a pseudo slashing, that should provide enough disincentive.

The pseudo slashing is implementing through the addition of an extra message with the proof for received invalid signatures, sent on the consensus. This message is not mandatory to be sent during a consensus round, but if it is sent, the proof will be verified and used by nodes listening to the consensus messages to temporarily blacklist the misbehaving nodes.
While blacklisted, the messages from the nodes in the blacklist will be dropped and this will cause rating drop and the eventual jailing, that penalize the node through the unjailing fee, missing block rewards/fees and queue time for becoming eligible for consensus again.

## Testing procedure
- Standard system test

- run with log trace for consensus
- Custom system test with invalid signers
     - there is a separate branch from this feat with invalid signer node `optimistic-consensus-sigcheck-malicious-node`
     - start a separate node in consensus with this branch
- TDB

### CODE REVIEW

File: consensus/spos/bls/export_test.go [hunk 1]
```diff
@@ -295,7 +299,27 @@ func (sr *subroundEndRound) IsOutOfTime() bool {
 	return sr.isOutOfTime()
 }
 
+func (sr *subroundEndRound) VerifyNodesOnAggSigVerificationFail() ([]string, error) {
```
Comment (SebastianMarian) [inline consensus/spos/bls/export_test.go:302]: VerifyNodesOnAggSigFail

File: factory/consensus/consensusComponents.go [hunk 2]
```diff
@@ -253,7 +259,9 @@ func (ccf *consensusComponentsFactory) Create() (*consensusComponents, error) {
 		FallbackHeaderValidator:       ccf.processComponents.FallbackHeaderValidator(),
 		NodeRedundancyHandler:         ccf.processComponents.NodeRedundancyHandler(),
 		ScheduledProcessor:            ccf.scheduledProcessor,
-		SignatureHandler:              signatureHandler,
+		MessageSigningHandler:         &mock.MessageSignerMock{},
```
[the remark this answers is not in the document]
Comment (ssd04) [reply factory/consensus/consensusComponents.go]: done

File: consensus/spos/bls/export_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/spos/bls/export_test.go]: done

**Comment by ssd04** [comment]
note to not forget about release tag for p2p repo

**Comment by mregolder** [comment]
How many nodes can be dropped because of blacklisting at any moment? Could the blacklisting mechanism be somehow exploited to trigger consecutive node drops? (I'm asking without really understanding how the consensus works)

**Comment by AdoAdoAdo** [comment]
> the mechanism shouldn't be susceptible to exploits, of this nature (to trigger node drops) when the nodes behave-follow the protocol, as the blacklisting relies on 2 signatures verification, where one of the signatures (p2p - ecdsa w/ secp256k1 verifies) but the BLS sig is invalid, which is clearly malicious behavior, and cannot happen when nodes follow the protocol.

As for how many nodes can be dropped (jailed due to the blacklisting) - there is no limit, as many as there are. Even if all nodes would be jailed, there is one protection where part of the jailed nodes are still able to participate in consensus in order to maintain/satisfy the 400 nodes/shard constraint

File: consensus/blacklist/blacklist.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	go pb.startSweepingPeerCache(ctx)
+}
+
+func (pb *peerBlacklist) startSweepingPeerCache(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("peerBlacklist's go routine is stopping...")
+			return
+		case <-time.After(durationSweepP2PBlacklist):
```
Comment (iulianpascalau) [inline consensus/blacklist/blacklist.go:92]: suggestion to replace the time.After with a timer.
Example: https://github.com/ElrondNetwork/elrond-go/blob/bd762dcc88845e11bc2ad51fe0d9af3dedbf37c8/common/statistics/resourceMonitor.go#L173

File: consensus/message.proto [hunk 4]
```diff
@@ -22,4 +22,5 @@ message Message {
 	bytes    AggregateSignature = 11;
 	bytes    LeaderSignature    = 12;
 	bytes    OriginatorPid      = 13;
+    bytes    InvalidSigners     = 14;
```
[the remark this answers is not in the document]
Comment (ssd04) [reply consensus/message.proto]: done

File: consensus/blacklist/blacklist.go (hunk 3 above)
Comment (ssd04) [reply consensus/blacklist/blacklist.go]: added a timer here

**Review by gabi-vuls — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**